File: nova/mobile/skills/base_skill.py
"""BaseSkill — Abstract base class for all Nova skills."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Optional
from nova.mobile.skills.manifest import SkillManifest
from nova.mobile.skills.context import SkillContext
from nova.mobile.skills.enums import SkillStatus


@dataclass
class SkillExecutionResult:
    skill_id: str
    is_success: bool
    spoken_response: str
    error_message: Optional[str] = None
    duration_ms: int = 0


class BaseSkill(ABC):
    @property
    @abstractmethod
    def manifest(self) -> SkillManifest:
        pass

    def __init__(self):
        self.context: Optional[SkillContext] = None
        self.status: SkillStatus = SkillStatus.NOT_INSTALLED

    def on_create(self): pass
    def on_enable(self): pass
    def on_disable(self): pass
    def on_destroy(self): pass

    @abstractmethod
    def can_handle(self, text: str) -> bool:
        pass

    @abstractmethod
    def execute(self, text: str) -> SkillExecutionResult:
        pass

File: mobile/skills/BaseSkill.kt
package com.nova.mobile.skills

data class SkillExecutionResult(
    val skillId: String,
    val isSuccess: Boolean,
    val spokenResponse: String,
    val errorMessage: String? = null,
    val durationMs: Long = 0L
)

/**
 * BaseSkill — Abstract base class all skills must extend.
 * Provides manifest, context, and a standardised execute() contract.
 */
abstract class BaseSkill {
    abstract val manifest: SkillManifest

    lateinit var context: SkillContext
        internal set

    var status: SkillStatus = SkillStatus.NOT_INSTALLED
        internal set

    /** Called once when the skill is loaded. */
    open fun onCreate() {}

    /** Called when the skill is enabled after being disabled. */
    open fun onEnable() {}

    /** Called when the skill is disabled. */
    open fun onDisable() {}

    /** Called before the skill is unloaded/removed. */
    open fun onDestroy() {}

    /** Returns true if this skill can handle the given text. */
    abstract fun canHandle(text: String): Boolean

    /** Execute the skill for the given input and return a spoken response. */
    abstract fun execute(text: String): SkillExecutionResult
}
